lightning/src/ln/channelmanager.rs

```diff
@@ -2589,6 +2594,64 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 		}
 	}
 
+	// Check that an inbound payment's `payment_data` field is sane. See `create_inbound_payment` and
+	// `create_inbound_payment_for_hash` for details on our checks.
+	fn verify_inbound_payment_data(&self, payment_hash: PaymentHash, payment_data: msgs::FinalOnionHopData) -> Result<Option<PaymentPreimage>, ()> {
+		let (iv_bytes, encrypted_metadata_bytes) = payment_data.payment_secret.0.split_at(16);
+		let mut chacha = ChaCha20::new(&self.inbound_payment_key[..], &iv_bytes[..12]);
```

Comment:
Right, we're gonna have to define our own custom variant of ChaCha here to use the full 16 byte seed. ChaCha20-RFC defaults to using 4 bytes for the counter and 12 bytes for the IV, but we actually only want one block and really want to use the full 16 bytes for the IV. If we drop the counter bytes we can replace them with the other 4 bytes of IV. This is algorithmically identical to using the 12 byte IV and then having 4 bytes of magic "seek to output index X", i.e. it can't be *worse* than simply using the 12-byte nonce, but it should actually provide 16 byte IV. API-wise, I think we should have a function in `ChaCha20` that takes a 32-byte key and 16-byte nonce and then simply returns 32/64 bytes from the ChaCha stream, communicating via API that this variant *only* supports one block.